### PR TITLE
#2128 download from pacs fails

### LINKED ISSUE #2128: NIFTI conversion fails (Missing start boundary)
For some datasets on OFSEP, the NIFTI conversion (with any converter) will fail.

The following logs are written in MS datasets :

```java
ofsep-prod-datasets-1  | 2024-03-05T11:03:59.632+01:00 ERROR 1 --- [nio-9904-exec-3] o.s.ng.download.WADODownloaderService    : A dicom file could not be downloaded from the pacs:
ofsep-prod-datasets-1  | 
ofsep-prod-datasets-1  | jakarta.mail.internet.ParseException: Missing start boundary
ofsep-prod-datasets-1  | 	at jakarta.mail.internet.MimeMultipart.parse(MimeMultipart.java:667)
ofsep-prod-datasets-1  | 	at jakarta.mail.internet.MimeMultipart.getCount(MimeMultipart.java:324)
ofsep-prod-datasets-1  | 	at org.shanoir.ng.download.WADODownloaderService.extractDICOMFilesFromMHTMLFile(WADODownloaderService.java:421)
ofsep-prod-datasets-1  | 	at org.shanoir.ng.download.WADODownloaderService.downloadDicomFilesForURLs(WADODownloaderService.java:282)
ofsep-prod-datasets-1  | 	at org.shanoir.ng.dataset.service.DatasetDownloaderServiceImpl.reconvertNifti(DatasetDownloaderServiceImpl.java:262)
ofsep-prod-datasets-1  | 	at org.shanoir.ng.dataset.service.DatasetDownloaderServiceImpl.massiveDownload(DatasetDownloaderServiceImpl.java:158)
ofsep-prod-datasets-1  | 	at org.shanoir.ng.dataset.controler.DatasetApiController.downloadDatasetById(DatasetApiController.java:330)
ofsep-prod-datasets-1  | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
ofsep-prod-datasets-1  | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
ofsep-prod-datasets-1  | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
ofsep-prod-datasets-1  | 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

The error occurs in `WADODownloaderService.extractDICOMFilesFromMHTMLFile(...)`, when `MimeMultipart.getCount()` is called.

These methods are only used when reconverting to NIFTI (see `DatasetDownloaderServiceImpl.reconvertNifti(...)`), and the mass download is OK for these DICOMs otherwise. Shouldn't we use the same method for all download, those for internal use included ?

### PR DESCRIPTION
**Pull request by jcomedouteau**
How to test:
-Find a dataset with a dataset file as WADO URI (containing /instances/) and anothter without.
- Download both datasets
- Both datasets can be downloaded

Closes #2128

### CODE REVIEW

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java [hunk 1]
```diff
@@ -159,7 +156,7 @@ public List<String> downloadDicomFilesForURLsAsZip(final List<URL> urls, final Z
 				String name = buildFileName(subjectName, dataset, datasetFilePath, sopInstanceUID);
 				// Download and zip
 				try {
-					String zipedFile = downloadAndWriteFileInZip(url, zipOutputStream, name);
+					String zipedFile = downloadAndWriteFileInZip(url, zipOutputStream, name, url.contains(WADO_REQUEST_TYPE_WADO_RS));
```
Comment (youennmerel) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java:159]: I think test should be directly made in downloadAndWriteFileInZip, as the tested url is in the parameters

```
// remove isMultipart parameter from method signature
private String downloadAndWriteFileInZip(String url, ZipOutputStream zipOutputStream, String name) throws ZipPacsFileException {
   ....
   this.extractDICOMZipFromMHTMLFile(responseBody,  name, zipOutputStream, url.contains(WADO_REQUEST_TYPE_WADO_RS));
   ...
}
```

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java [hunk 2]
```diff
@@ -425,9 +422,16 @@ private void extractDICOMFilesFromMHTMLFile(final byte[] responseBody, final Str
 	 * @throws IOException
 	 * @throws MessagingException
 	 */
-	private void extractDICOMZipFromMHTMLFile(final byte[] responseBody, String name, ZipOutputStream zipOutputStream)
+	private void extractDICOMZipFromMHTMLFile(final byte[] responseBody, String name, ZipOutputStream zipOutputStream, boolean isMultipart)
 			throws IOException, MessagingException {
 		try(ByteArrayInputStream bIS = new ByteArrayInputStream(responseBody)) {
+			if (!isMultipart) {
```
Comment (youennmerel) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java:428]: I think this method should be now structured like this :

```
// not multipart
if (!isMultipart) {
   ...
   return;
}

...
// multipart with 1 file
if (count == 1) {
   // invert condition here
   if (!(bodyPart.isMimeType(CONTENT_TYPE_DICOM) && bodyPart.isMimeType(CONTENT_TYPE_DICOM_XML))) {
      throw new IOException(...);
   }
   ...
   return
}

// multipart with multiple file
for (int i = 0; i < count; i++) {
   // invert condition here
   if (!(bodyPart.isMimeType(CONTENT_TYPE_DICOM) && bodyPart.isMimeType(CONTENT_TYPE_DICOM_XML))) {
      throw new IOException(...);
   }
   ...
}
```

Would clarify cases, and eliminate a bit of cyclomatic complexity

**Comment by michaelkain** [comment]
Excellent!

**Review by youennmerel — APPROVED** [review]
Tested on neurinfo qualif, OK for me !

**Pull request merged**